test/services/GasEstimator.test.ts

```diff
@@ -0,0 +1,233 @@
+import { expect } from 'chai';
+import sinon, { StubbedInstance, stubConstructor } from 'ts-sinon';
+import { BigNumber, ethers } from 'ethers';
+
+import { GasEstimator } from '../../src/services/GasEstimator';
+import { GasEstimation } from '../../src/types/GasEstimation';
+import { BlockWithTransactions, FeeData, TransactionResponse } from '@ethersproject/abstract-provider';
+
+describe('GasEstimator', () => {
+  describe('.apply', () => {
+    const fakeProvider: StubbedInstance<ethers.providers.BaseProvider> = stubConstructor(ethers.providers.BaseProvider);
+
+    beforeEach(() => {
+      fakeProvider.getGasPrice.resolves();
+      fakeProvider.getBlockWithTransactions.resolves();
+      fakeProvider.getFeeData.resolves();
+    });
+
+    afterEach(() => {
+      sinon.restore();
+    });
+
+    before(() => {
+      GasEstimator.apply(fakeProvider, 1e6, 1e9).then();
+    });
+
+    it('calls getGasPrice from Provider', () => {
+      expect(fakeProvider.getGasPrice.called).to.be.true;
+    });
+
+    it('calls getBlockWithTransaction from Provider', () => {
+      expect(fakeProvider.getBlockWithTransactions.called).to.be.true;
+    });
+
+    it('calls getFeeData from Provider', () => {
+      expect(fakeProvider.getFeeData.called).to.be.true;
+    });
+
+    describe('when provider fails to retrieve data', () => {
+      beforeEach(() => {
+        fakeProvider.getGasPrice.rejects();
+        fakeProvider.getBlockWithTransactions.rejects();
+        fakeProvider.getFeeData.rejects();
+      });
+
+      afterEach(() => {
+        sinon.restore();
+      });
+
+      it('throws an error', () => {
+        expect(GasEstimator.apply(fakeProvider, 1e6, 1e9)).to.eventually.throw();
+      });
+    });
+
+    describe('when params are inserted correctly', () => {
+      beforeEach(() => {
+        fakeProvider.getGasPrice.resolves(BigNumber.from(100));
+        fakeProvider.getBlockWithTransactions.resolves({
+          transactions: [
+            {
+              gasPrice: BigNumber.from(100),
+              maxPriorityFeePerGas: BigNumber.from(100),
+            } as TransactionResponse,
+          ],
+        } as BlockWithTransactions);
+        fakeProvider.getFeeData.resolves(<FeeData>{ maxPriorityFeePerGas: BigNumber.from(100) });
+      });
+
+      afterEach(() => {
```

Comment:
I think you don't need this afterEach since it is already defined at L19

Reply:
I'll remove it and test again. Thanks!